allennlp/data/data_loaders/data_collator.py

```diff
@@ -0,0 +1,68 @@
+from typing import List
+
+from transformers.data.data_collator import DataCollatorForLanguageModeling
+from allennlp.common import Registrable
+from allennlp.data.batch import Batch
+from allennlp.data.data_loaders.data_loader import TensorDict
+from allennlp.data.instance import Instance
+
+
+def allennlp_collate(instances: List[Instance]) -> TensorDict:
+    """
+    This is the default function used to turn a list of `Instance`s into a `TensorDict`
+    batch.
+    """
+    batch = Batch(instances)
+    return batch.as_tensor_dict()
+
+
+class DataCollator(Registrable):
+    """
+    This class is similar with `DataCollator` in [Transformers]
+    (https://github.com/huggingface/transformers/blob/master/src/transformers/data/data_collator.py)
+    Allow to do some dynamic operations for tensor in different batches
+    Cause this method run before each epoch to convert `List[Instance]` to `TensorDict`
+    """
+
+    def __call__(self, instances: List[Instance]) -> TensorDict:
+        raise NotImplementedError
+
+
+class DefaultDataCollator(DataCollator):
+    def __call__(self, instances: List[Instance]) -> TensorDict:
+        return allennlp_collate(instances)
+
+
+@DataCollator.register("language-model")
```

Comment:
```suggestion
@DataCollator.register("language_model")
```

We should be using underscores for everything. It's annoying when you have to check which one it is while writing a config file.

The only exceptions I can think of are commands and vocabularies. For commands, I kind of get it. Commands on the command line usually use dashes. For vocabularies we also use dashes, which I think is wrong, but I don't want to break backwards compatibility for it.